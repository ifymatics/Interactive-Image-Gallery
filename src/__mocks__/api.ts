export const getMe = jest.fn();
export const login = jest.fn();
export const register = jest.fn();
export const logout = jest.fn();
