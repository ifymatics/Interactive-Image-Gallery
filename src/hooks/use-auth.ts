"use client";

import useSWR, { useSWRConfig } from "swr";
import {
  getMe,
  login as apiLogin,
  register as apiRegister,
  logout as apiLogout,
} from "@/lib/api";
import type { User, AuthCredentials } from "@/types";
import { useCallback } from "react";

export function useAuth() {
  const {
    data: user,
    error,
    isLoading,
    mutate,
  } = useSWR<User | null>("/auth/me", getMe, {
    shouldRetryOnError: false, // Don't retry on 401/403
  });
  const { cache } = useSWRConfig();

  const login = useCallback(
    async (credentials: AuthCredentials) => {
      try {
        await apiLogin(credentials);
        await mutate(); // Re-fetches /auth/me to get user
      } catch (err) {
        console.error("Login failed", err);
        throw err; // Re-throw to be caught in the form
      }
    },
    [mutate]
  );

  const signup = useCallback(
    async (credentials: AuthCredentials) => {
      try {
        await apiRegister(credentials);
        await mutate(); // Re-fetches /auth/me to get user
      } catch (err) {
        console.error("Signup failed", err);
        throw err; // Re-throw to be caught in the form
      }
    },
    [mutate]
  );

  // const logout = useCallback(async () => {
  //   try {
  //     await apiLogout();
  //     // @ts-ignore
  //     cache.clear();
  //     await mutate(null, false); // Set user to null without revalidation
  //   } catch (err) {
  //     console.error("Logout failed", err);
  //   }
  // }, [mutate, cache]);
  const logout = useCallback(async () => {
    try {
      await apiLogout();

      // âœ… Check if cache has a clear function (like Map)
      if (typeof (cache as any).clear === "function") {
        // (cache as Map<any, any>).clear();
      }

      await mutate(null, false);
    } catch (err) {
      console.error("Logout failed", err);
      await mutate(null, false); // Still remove user
    }
  }, [mutate, cache]);

  return {
    user: user ?? null,
    isLoading,
    isError: error,
    login,
    signup,
    logout,
  };
}
