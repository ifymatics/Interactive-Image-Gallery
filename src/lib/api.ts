import axios from "axios";
import type {
  UnsplashApiResponse,
  Comment,
  AuthCredentials,
  User,
  LikeCount,
  HasLiked,
} from "@/types";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:5000";
const UNSPLASH_API_URL = "https://api.unsplash.com";
const UNSPLASH_ACCESS_KEY = process.env.NEXT_PUBLIC_UNSPLASH_ACCESS_KEY;

const api = axios.create({
  baseURL: API_URL,
  withCredentials: true,
});

const unsplashApi = axios.create({
  baseURL: UNSPLASH_API_URL,
  headers: {
    Authorization: `Client-ID ${UNSPLASH_ACCESS_KEY}`,
  },
});

// --- Image Fetching ---
export const fetchImages = async (
  page: number,
  perPage: number
): Promise<UnsplashApiResponse> => {
  const response = await unsplashApi.get("/search/photos", {
    params: {
      query: "nature",
      page,
      per_page: perPage,
    },
  });
  return response.data;
};

// --- Authentication ---
export const getMe = async (): Promise<User | null> => {
  try {
    const response = await api.get("/auth/me");
    return response.data;
  } catch (error) {
    return null;
  }
};

export const register = async (credentials: AuthCredentials): Promise<void> => {
  await api.post("/auth/register", credentials);
};

export const login = async (credentials: AuthCredentials): Promise<void> => {
  await api.post("/auth/login", credentials);
};

export const logout = async (): Promise<void> => {
  await api.post("/auth/logout");
};

// --- Comments ---
export const fetchComments = async (imageId: string): Promise<Comment[]> => {
  const response = await api.get(`/comments`, { params: { imageId } });
  return response.data;
};

export const postComment = async (data: {
  imageId: string;
  text: string;
}): Promise<Comment> => {
  const response = await api.post("/comments", data);
  return response.data;
};

// --- Likes ---
export const fetchLikeCount = async (imageId: string): Promise<LikeCount> => {
  const response = await api.get("/likes/count", { params: { imageId } });
  return response.data;
};

export const hasLiked = async (imageId: string): Promise<HasLiked> => {
  const response = await api.get("/likes/has-liked", { params: { imageId } });
  return response.data;
};

export const likeImage = async (imageId: string): Promise<void> => {
  await api.post("/likes", null, { params: { imageId } });
};

export const unlikeImage = async (imageId: string): Promise<void> => {
  await api.delete("/likes", { params: { imageId } });
};
