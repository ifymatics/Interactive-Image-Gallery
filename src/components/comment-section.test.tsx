// src/components/comment-section.test.tsx
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CommentSection from "@/components/comment-section";
import { useAuth } from "@/hooks/use-auth";
import useSWR, { MutatorOptions } from "swr";
import { postComment } from "@/lib/api";

// Mock dependencies
jest.mock("@/hooks/use-auth");
jest.mock("swr");
jest.mock("@/lib/api");
jest.mock("@/components/ui/avatar", () => ({
  Avatar: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  AvatarFallback: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));
jest.mock("@/components/ui/skeleton", () => ({
  Skeleton: ({ className }: { className?: string }) => (
    <div className={className}>Skeleton</div>
  ),
}));
jest.mock("lucide-react", () => ({
  MessageSquare: () => <div>MessageIcon</div>,
}));
jest.mock("./comment-form", () => ({
  __esModule: true,
  default: ({
    onSubmit,
    isDisabled,
  }: {
    onSubmit: (text: string) => void;
    isDisabled: boolean;
  }) => (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        onSubmit("Test comment");
      }}
    >
      <button type="submit" disabled={isDisabled}>
        Submit Comment
      </button>
    </form>
  ),
}));

const mockUseAuth = useAuth as jest.MockedFunction<typeof useAuth>;
const mockUseSWR = useSWR as jest.MockedFunction<typeof useSWR>;
const mockPostComment = postComment as jest.MockedFunction<typeof postComment>;

describe("CommentSection", () => {
  const mockImageId = "test-image-id";
  const mockOnAuthRequired = jest.fn();
  const mockMutate = jest.fn();

  const mockComments = [
    {
      id: "1",
      imageId: mockImageId,
      user: { username: "user1" },
      text: "First comment",
      createdAt: new Date().toISOString(),
    },
    {
      id: "2",
      imageId: mockImageId,
      user: { username: "user2" },
      text: "Second comment",
      createdAt: new Date(Date.now() - 100000).toISOString(),
    },
  ];

  beforeEach(() => {
    mockUseAuth.mockReturnValue({
      user: null,
      login: jest.fn(),
      signup: jest.fn(),
      logout: jest.fn(),
      isLoading: false,
      isError: undefined,
    });

    mockUseSWR.mockReturnValue({
      data: mockComments,
      mutate: mockMutate,
      isLoading: false,
      error: undefined,
      isValidating: false,
    });

    mockPostComment.mockResolvedValue({
      id: "new-comment-id",
      imageId: mockImageId,
      user: {
        username: "testuser",
        id: "",
      },
      text: "Test comment",
      createdAt: new Date().toISOString(),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders loading state correctly", () => {
    mockUseSWR.mockReturnValue({
      isLoading: true,
      data: undefined,
      error: undefined,
      mutate: function <MutationData = unknown>(
        data?: unknown,
        opts?: boolean | MutatorOptions<unknown, MutationData> | undefined
      ): Promise<unknown> {
        throw new Error("Function not implemented.");
      },
      isValidating: false,
    });
    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={false}
      />
    );

    expect(screen.getAllByText("Skeleton").length).toBe(2);
  });

  it("renders empty state when no comments exist", () => {
    mockUseSWR.mockReturnValue({
      data: [],
      isLoading: false,
      error: undefined,
      mutate: function <MutationData = unknown>(
        data?: unknown,
        opts?: boolean | MutatorOptions<unknown, MutationData> | undefined
      ): Promise<unknown> {
        throw new Error("Function not implemented.");
      },
      isValidating: false,
    });
    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={false}
      />
    );

    expect(
      screen.getByText("Be the first to leave a comment.")
    ).toBeInTheDocument();
  });

  it("displays comments correctly", () => {
    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={false}
      />
    );

    expect(screen.getByText("First comment")).toBeInTheDocument();
    expect(screen.getByText("Second comment")).toBeInTheDocument();
    expect(screen.getByText("user1")).toBeInTheDocument();
    expect(screen.getByText("user2")).toBeInTheDocument();
  });

  it("shows auth required when submitting without being logged in", async () => {
    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={false}
      />
    );

    const submitButton = screen.getByText("Submit Comment");
    await userEvent.click(submitButton);

    expect(mockOnAuthRequired).toHaveBeenCalled();
    expect(mockPostComment).not.toHaveBeenCalled();
  });

  it("handles successful comment submission with optimistic update", async () => {
    const mockUser = { username: "testuser", id: "4" };
    mockUseAuth.mockReturnValue({
      user: mockUser,
      login: jest.fn(),
      signup: jest.fn(),
      logout: jest.fn(),
      isLoading: false,
      isError: undefined,
    });

    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={false}
      />
    );

    const submitButton = screen.getByText("Submit Comment");
    await userEvent.click(submitButton);

    // Check optimistic update
    expect(mockMutate).toHaveBeenCalledWith(
      expect.arrayContaining([
        expect.objectContaining({
          id: expect.any(String),
          imageId: mockImageId,
          user: mockUser,
          text: "Test comment",
          createdAt: expect.any(String),
        }),
        ...mockComments,
      ]),
      false
    );

    // Check final update after API call
    await waitFor(() => {
      expect(mockPostComment).toHaveBeenCalledWith({
        imageId: mockImageId,
        text: "Test comment",
      });
      expect(mockMutate).toHaveBeenCalledTimes(2); // Once for optimistic, once for revalidation
    });
  });

  it("handles comment submission error and reverts optimistic update", async () => {
    const mockUser = { username: "testuser", id: "1" };
    mockUseAuth.mockReturnValue({
      user: mockUser,
      login: jest.fn(),
      signup: jest.fn(),
      logout: jest.fn(),
      isLoading: false,
      isError: undefined,
    });

    const testError = new Error("Failed to post comment");
    mockPostComment.mockRejectedValueOnce(testError);

    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={false}
      />
    );

    const submitButton = screen.getByText("Submit Comment");
    await userEvent.click(submitButton);

    // Check optimistic update
    expect(mockMutate).toHaveBeenCalledWith(
      expect.arrayContaining([
        expect.objectContaining({
          text: "Test comment",
        }),
        ...mockComments,
      ]),
      false
    );

    // Check error handling
    await waitFor(() => {
      expect(mockPostComment).toHaveBeenCalled();
      expect(mockMutate).toHaveBeenCalledTimes(2); // Once for optimistic, once for revalidation
    });
  });

  it("disables form when auth is loading", () => {
    mockUseAuth.mockReturnValue({
      user: null,
      login: jest.fn(),
      signup: jest.fn(),
      logout: jest.fn(),
      isLoading: true,
      isError: undefined,
    });

    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={true}
      />
    );

    const submitButton = screen.getByText("Submit Comment");
    expect(submitButton).toBeDisabled();
  });

  it("filters comments by imageId", () => {
    const mixedComments = [
      ...mockComments,
      {
        id: "3",
        imageId: "other-image-id",
        user: { username: "user3" },
        text: "Comment for other image",
        createdAt: new Date().toISOString(),
      },
    ];

    mockUseSWR.mockReturnValue({
      data: mixedComments,
      isLoading: false,
      error: undefined,
      mutate: function <MutationData = unknown>(
        data?: unknown,
        opts?: boolean | MutatorOptions<unknown, MutationData> | undefined
      ): Promise<unknown> {
        throw new Error("Function not implemented.");
      },
      isValidating: false,
    });

    render(
      <CommentSection
        imageId={mockImageId}
        onAuthRequired={mockOnAuthRequired}
        isAuthLoading={false}
      />
    );

    expect(screen.getByText("First comment")).toBeInTheDocument();
    expect(screen.getByText("Second comment")).toBeInTheDocument();
    expect(
      screen.queryByText("Comment for other image")
    ).not.toBeInTheDocument();
  });
});
