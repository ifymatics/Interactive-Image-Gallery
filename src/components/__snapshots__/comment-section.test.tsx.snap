// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentSkeleton renders correctly 1`] = `
<div
  class="flex items-start space-x-3"
>
  <div
    class="h-8 w-8 rounded-full bg-gray-200"
  />
  <div
    class="flex-1 space-y-2"
  >
    <div
      class="h-4 w-1/2 bg-gray-200"
    />
    <div
      class="h-4 w-full bg-gray-200"
    />
  </div>
</div>
`;
